subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c

```diff
@@ -704,6 +719,104 @@ gst_msdk_export_dmabuf_to_vasurface (GstMsdkContext * context,
   }
 }
 
+static VASurfaceID
+_get_va_surface (GstBuffer * buf, GstVideoInfo * info,
+    GstMsdkContext * msdk_context)
+{
+  VASurfaceID va_surface = VA_INVALID_ID;
+
+  if (info == NULL) {
+    va_surface = gst_va_buffer_get_surface (buf);
+  } else {
+    GstVideoMeta *vmeta;
+    GstMemory *mem;
+    guint i, fd;
+
+    mem = gst_buffer_peek_memory (buf, 0);
+    fd = gst_dmabuf_memory_get_fd (mem);
+    if (fd < 0)
+      return va_surface;
+
+    vmeta = gst_buffer_get_video_meta (buf);
+    if (vmeta) {
+      if (GST_VIDEO_INFO_FORMAT (info) != vmeta->format ||
+          GST_VIDEO_INFO_WIDTH (info) != vmeta->width ||
+          GST_VIDEO_INFO_HEIGHT (info) != vmeta->height ||
+          GST_VIDEO_INFO_N_PLANES (info) != vmeta->n_planes) {
+        GST_ERROR ("VideoMeta attached to buffer is not matching"
+            "the negotiated width/height/format");
+        return va_surface;
+      }
+      for (i = 0; i < GST_VIDEO_INFO_N_PLANES (info); ++i) {
+        GST_VIDEO_INFO_PLANE_OFFSET (info, i) = vmeta->offset[i];
+        GST_VIDEO_INFO_PLANE_STRIDE (info, i) = vmeta->stride[i];
+      }
+      GST_VIDEO_INFO_SIZE (info) = gst_buffer_get_size (buf);
+    }
+    /* export dmabuf to vasurface */
+    if (!gst_msdk_export_dmabuf_to_vasurface (msdk_context, info, fd,
+            &va_surface)) {
+      return VA_INVALID_ID;
+    }
+  }
+
+  return va_surface;
+}
+
+GstMsdkSurface *
+import_to_msdk_surface (GstBuffer * buf, GstMsdkContext * msdk_context,
+    GstVideoInfo * info)
+{
+  VASurfaceID va_surface = VA_INVALID_ID;
+  GstMemory *mem = NULL;
+  mfxFrameInfo frame_info = { 0, };
+  GstMsdkSurface *msdk_surface = NULL;
+  mfxFrameSurface1 *mfx_surface = NULL;
+  GstMsdkMemoryID *msdk_mid = NULL;
+  mfxMemId *mfx_mid = NULL;
+
+  mem = gst_buffer_peek_memory (buf, 0);
+  msdk_surface = g_slice_new0 (GstMsdkSurface);
+
+  /* If buffer has qdata pointing to mfxFrameSurface1, directly extract it */
+  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
+              GST_MSDK_FRAME_SURFACE))) {
+    msdk_surface->surface = mfx_surface;
+    msdk_surface->from_qdata = TRUE;
+    return msdk_surface;
+  }
+
+  if (gst_msdk_is_va_mem (mem)) {
+    va_surface = _get_va_surface (buf, NULL, NULL);
+  } else if (gst_is_dmabuf_memory (mem)) {
+    va_surface = _get_va_surface (buf, info, msdk_context);
+  }
+
+  if (va_surface == VA_INVALID_ID)
+    return NULL;
+
+  mfx_surface = g_slice_new0 (mfxFrameSurface1);
+  msdk_mid = g_slice_new0 (GstMsdkMemoryID);
+  mfx_mid = g_slice_new0 (mfxMemId);
+
+  msdk_mid->surface = g_slice_new0 (VASurfaceID);
+  *msdk_mid->surface = va_surface;
```

Comment:
I am wondering why we need to use malloc here? It just contains one surface and no need to use stack.

Reply:
I think GstMsdkMemoryID was designed to save a bunch of VASurfaceIDs, the original code used vaCreateSurfaces API  to obtain request->NumFrameSuggested va surfaces. In refact code, since we want to use va pool use allocate gstbuffer with va memory, so it would be one va surface for one gstbuffer, and there is no need to use a pointer *VASurfaceIDs in GstMsdkMemoryID. However, considering the decoder modified codes are not included here, directly changing GstMsdkMmemoryID will have an impact on decoder (mainly the mfxFrameAllocator callback functions). I would prefer not to change this part here until we move to decoder code.

Reply:
OK.